File: src/voidx/skills/registry.py
"""Local skill discovery from SKILL.md files."""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

from voidx.platform.paths import voidx_global_skills_dir, voidx_workspace_skills_dir
from voidx.skills.schema import SkillDefinition, SkillMeta, SkillScope

SKILL_FILENAME = "SKILL.md"
DEFAULT_BUNDLED_DIR = Path(__file__).resolve().parent / "bundled"
SKILL_NAME_RE = re.compile(r"^(?=.{1,64}$)[a-z0-9]([a-z0-9-]*[a-z0-9])?$")


@dataclass
class _ParsedSkillCacheEntry:
    mtime_ns: int
    size: int
    skill: SkillDefinition


_PARSED_SKILL_CACHE: dict[tuple[str, SkillScope], _ParsedSkillCacheEntry] = {}


def normalize_skill_name(name: str) -> str:
    return name.strip().lower()


class SkillParseError(ValueError):
    pass


class SkillRegistry:
    """Discovers bundled, global, and project skills.

    Search order is bundled first, then global, then project. Later sources
    override earlier sources with the same skill name.
    """

    def __init__(
        self,
        workspace: str,
        *,
        bundled_dir: Path | None = None,
        global_dir: Path | None = None,
        project_dir: Path | None = None,
    ) -> None:
        self.workspace = Path(workspace).resolve()
        self.bundled_dir = bundled_dir or DEFAULT_BUNDLED_DIR
        self.global_dir = global_dir or voidx_global_skills_dir()
        self.project_dir = project_dir or voidx_workspace_skills_dir(self.workspace)
        self._cache: list[SkillDefinition] | None = None
        self._cache_signature: tuple[tuple[str, str, int, int], ...] | None = None

    def discover(self) -> list[SkillDefinition]:
        signature = self._discover_signature()
        if self._cache is not None and self._cache_signature == signature:
            return self._cache
        skills: dict[str, SkillDefinition] = {}
        for scope, root in (
            ("bundled", self.bundled_dir),
            ("global", self.global_dir),
            ("project", self.project_dir),
        ):
            for skill in self._discover_root(root, scope):
                skills[normalize_skill_name(skill.name)] = skill
        self._cache = sorted(skills.values(), key=lambda item: item.name)
        self._cache_signature = signature
        return self._cache

    def invalidate(self) -> None:
        self._cache = None
        self._cache_signature = None

    def create_skill(
        self,
        name: str,
        description: str,
        body: str,
        *,
        scope: Literal["project", "global"] = "project",
    ) -> Path | None:
        """Create a SKILL.md file. Returns the path, or None if it already exists.

        Security: name is validated against SKILL_NAME_RE on the first line.
        This is the sole path-escape defense since sandbox skips the check
        (skill create has no file_path arg). Must not rely on tool-layer validation.
        """
        if not SKILL_NAME_RE.match(name):
            raise ValueError(
                f"Invalid skill name '{name}': must be 1-64 chars, lowercase "
                f"alphanumeric with hyphens, not starting/ending with a hyphen."
            )

        root = self.project_dir if scope == "project" else self.global_dir
        path = root / name / SKILL_FILENAME
        if path.exists():
            return None

        path.parent.mkdir(parents=True, exist_ok=True)
        frontmatter = (
            "---\n"
            f"name: {name}\n"
            f"description: {description}\n"
            "enabled: true\n"
            "---\n\n"
        )
        path.write_text(frontmatter + body, encoding="utf-8")
        self.invalidate()
        return path


    def get(self, name: str) -> SkillDefinition | None:
        target = normalize_skill_name(name)
        for skill in self.discover():
            if normalize_skill_name(skill.name) == target:
                return skill
        return None

    def _discover_root(self, root: Path, scope: SkillScope) -> list[SkillDefinition]:
        if not root.exists() or not root.is_dir():
            return []
        skills: list[SkillDefinition] = []
        for skill_file in sorted(root.glob(f"*/{SKILL_FILENAME}")):
            try:
                skills.append(parse_skill_file(skill_file, scope=scope))
            except SkillParseError:
                continue
        return skills

    def _discover_signature(self) -> tuple[tuple[str, str, int, int], ...]:
        entries: list[tuple[str, str, int, int]] = []
        for scope, root in (
            ("bundled", self.bundled_dir),
            ("global", self.global_dir),
            ("project", self.project_dir),
        ):
            if not root.exists() or not root.is_dir():
                continue
            for skill_file in sorted(root.glob(f"*/{SKILL_FILENAME}")):
                try:
                    stat = skill_file.stat()
                except OSError:
                    continue
                entries.append((scope, str(skill_file.resolve()), stat.st_mtime_ns, stat.st_size))
        return tuple(entries)


def parse_skill_file(path: Path, *, scope: SkillScope) -> SkillDefinition:
    resolved = path.resolve()
    stat = resolved.stat()
    key = (str(resolved), scope)
    cached = _PARSED_SKILL_CACHE.get(key)
    if cached is not None and cached.mtime_ns == stat.st_mtime_ns and cached.size == stat.st_size:
        return cached.skill.model_copy(deep=True)

    text = resolved.read_text(encoding="utf-8", errors="replace")
    fields, body = _split_frontmatter(text)
    name = str(fields.get("name") or path.parent.name).strip()
    if not name:
        raise SkillParseError(f"Skill at {path} has no name")
    meta = SkillMeta(
        name=name,
        description=str(fields.get("description") or ""),
        enabled=_coerce_bool(fields.get("enabled"), default=True),
        triggers=[str(item).strip() for item in _coerce_list(fields.get("triggers")) if str(item).strip()],
        scope=scope,
    )
    skill = SkillDefinition(meta=meta, path=resolved, body=body.strip())
    _PARSED_SKILL_CACHE[key] = _ParsedSkillCacheEntry(
        mtime_ns=stat.st_mtime_ns,
        size=stat.st_size,
        skill=skill,
    )
    return skill.model_copy(deep=True)


def _split_frontmatter(text: str) -> tuple[dict[str, Any], str]:
    lines = text.splitlines()
    if not lines or lines[0].strip() != "---":
        return {}, text

    end_index = None
    for index, line in enumerate(lines[1:], start=1):
        if line.strip() == "---":
            end_index = index
            break
    if end_index is None:
        raise SkillParseError("Unclosed frontmatter")

    frontmatter = "\n".join(lines[1:end_index])
    body = "\n".join(lines[end_index + 1:])
    return _parse_frontmatter(frontmatter), body


def _parse_frontmatter(text: str) -> dict[str, Any]:
    result: dict[str, Any] = {}
    lines = text.splitlines()
    index = 0
    while index < len(lines):
        line = lines[index].rstrip()
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            index += 1
            continue
        if stripped.startswith("- ") and result:
            last_key = next(reversed(result))
            current = result[last_key]
            if isinstance(current, list):
                current.append(_parse_scalar(stripped[2:].strip()))
            index += 1
            continue
        if ":" not in line:
            index += 1
            continue
        key, value = line.split(":", 1)
        key = key.strip()
        value = value.strip()
        if not key:
            index += 1
            continue
        if value in ("", ">", "|"):
            collected, index = _collect_block(lines, index + 1, folded=(value == ">"))
            if value == "" and not collected:
                result[key] = []
            elif value == "":
                result[key] = collected[0] if len(collected) == 1 else collected
            else:
                result[key] = collected
        else:
            result[key] = _parse_scalar(value)
            index += 1
    return result


def _collect_block(lines: list[str], start: int, *, folded: bool = False) -> tuple[str, int]:
    parts: list[str] = []
    index = start
    while index < len(lines):
        line = lines[index]
        if not line.startswith(" ") and not line.startswith("\t") and line.strip():
            break
        stripped = line.strip()
        if stripped.startswith("- "):
            break
        if not stripped:
            index += 1
            continue
        parts.append(stripped)
        index += 1
    text = " ".join(parts) if folded else "\n".join(parts)
    return text, index


def _parse_scalar(value: str) -> Any:
    value = value.strip()
    if len(value) >= 2 and value[0] == value[-1] and value[0] in ("'", '"'):
        value = value[1:-1]
    lower = value.lower()
    if lower in {"true", "yes", "on"}:
        return True
    if lower in {"false", "no", "off"}:
        return False
    if value.startswith("[") and value.endswith("]"):
        inner = value[1:-1].strip()
        if not inner:
            return []
        return [_parse_scalar(part.strip()) for part in inner.split(",")]
    return value


def _coerce_bool(value: Any, *, default: bool) -> bool:
    if value is None:
        return default
    if isinstance(value, bool):
        return value
    return str(value).strip().lower() not in {"false", "no", "off", "0"}


def _coerce_list(value: Any) -> list[Any]:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    return [value]

File: crates/voidx-skills/src/registry.rs
//! Local skill discovery from SKILL.md files.
//!
//! Ported from `src/voidx/skills/registry.py`.
//!
//! Search order: bundled → global → project. Later sources override
//! earlier sources with the same skill name (case-insensitive).

use crate::schema::{SkillDefinition, SkillMeta, SkillScope};
use std::collections::HashMap;
use std::path::{Path, PathBuf};

const SKILL_FILENAME: &str = "SKILL.md";

/// Normalize a skill name for case-insensitive comparison.
pub fn normalize_skill_name(name: &str) -> String {
    name.trim().to_lowercase()
}

/// Error during skill file parsing.
#[derive(Debug, thiserror::Error)]
pub enum SkillParseError {
    #[error("Skill at {path} has no name")]
    NoName { path: PathBuf },
    #[error("Unclosed frontmatter in {path}")]
    UnclosedFrontmatter { path: PathBuf },
    #[error("IO error reading {path}: {source}")]
    Io {
        path: PathBuf,
        #[source]
        source: std::io::Error,
    },
}

/// Discovers bundled, global, and project skills.
pub struct SkillRegistry {
    workspace: PathBuf,
    bundled_dir: PathBuf,
    global_dir: PathBuf,
    project_dir: PathBuf,
    cache: Option<Vec<SkillDefinition>>,
}

impl SkillRegistry {
    pub fn new(workspace: &Path) -> Self {
        let workspace = workspace.to_path_buf().canonicalize().unwrap_or_else(|_| workspace.to_path_buf());
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/"));

        Self {
            workspace: workspace.clone(),
            bundled_dir: PathBuf::from(env!("CARGO_MANIFEST_DIR"))
                .join("bundled")
                .join("superpowers"),
            global_dir: home.join(".voidx").join("skills"),
            project_dir: workspace.join(".voidx").join("skills"),
            cache: None,
        }
    }

    /// Create with custom directories (for testing).
    pub fn with_dirs(
        workspace: PathBuf,
        bundled_dir: PathBuf,
        global_dir: PathBuf,
        project_dir: PathBuf,
    ) -> Self {
        Self {
            workspace,
            bundled_dir,
            global_dir,
            project_dir,
            cache: None,
        }
    }

    /// Discover all skills. Results are cached until `invalidate()` is called.
    pub fn discover(&mut self) -> &[SkillDefinition] {
        if self.cache.is_none() {
            let mut skills: HashMap<String, SkillDefinition> = HashMap::new();

            for (scope, root) in [
                (SkillScope::Bundled, &self.bundled_dir),
                (SkillScope::Global, &self.global_dir),
                (SkillScope::Project, &self.project_dir),
            ] {
                for skill in discover_root(root, scope) {
                    skills.insert(normalize_skill_name(skill.name()), skill);
                }
            }

            let mut list: Vec<SkillDefinition> = skills.into_values().collect();
            list.sort_by(|a, b| a.name().cmp(b.name()));
            self.cache = Some(list);
        }
        self.cache.as_ref().unwrap()
    }

    /// Invalidate the cache, forcing re-discovery on next `discover()`.
    pub fn invalidate(&mut self) {
        self.cache = None;
    }

    /// Get a specific skill by name.
    pub fn get(&mut self, name: &str) -> Option<&SkillDefinition> {
        let target = normalize_skill_name(name);
        self.discover();
        self.cache
            .as_ref()
            .unwrap()
            .iter()
            .find(|s| normalize_skill_name(s.name()) == target)
    }
}

/// Discover skills under a root directory.
fn discover_root(root: &Path, scope: SkillScope) -> Vec<SkillDefinition> {
    if !root.exists() || !root.is_dir() {
        return Vec::new();
    }

    let mut skills = Vec::new();

    // Look for */SKILL.md pattern
    if let Ok(entries) = std::fs::read_dir(root) {
        let mut dirs: Vec<_> = entries.filter_map(|e| e.ok()).collect();
        dirs.sort_by_key(|e| e.file_name());

        for entry in dirs {
            if !entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                continue;
            }
            let skill_file = entry.path().join(SKILL_FILENAME);
            if skill_file.exists() {
                match parse_skill_file(&skill_file, scope) {
                    Ok(skill) => skills.push(skill),
                    Err(e) => {
                        tracing::warn!("Failed to parse skill file {:?}: {}", skill_file, e);
                    }
                }
            }
        }
    }

    skills
}

/// Parse a SKILL.md file with optional YAML frontmatter.
pub fn parse_skill_file(path: &Path, scope: SkillScope) -> Result<SkillDefinition, SkillParseError> {
    let text = std::fs::read_to_string(path).map_err(|e| SkillParseError::Io {
        path: path.to_path_buf(),
        source: e,
    })?;

    let (fields, body) = split_frontmatter(&text);

    let name = fields
        .get("name")
        .and_then(|v| v.as_str())
        .map(|s| s.trim().to_string())
        .unwrap_or_else(|| {
            path.parent()
                .and_then(|p| p.file_name())
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default()
        });

    if name.is_empty() {
        return Err(SkillParseError::NoName {
            path: path.to_path_buf(),
        });
    }

    let description = fields
        .get("description")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let enabled = fields
        .get("enabled")
        .and_then(|v| v.as_bool())
        .unwrap_or(true);

    let triggers = fields
        .get("triggers")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(|s| s.trim().to_string()))
                .filter(|s| !s.is_empty())
                .collect()
        })
        .unwrap_or_default();

    let meta = SkillMeta {
        name,
        description,
        enabled,
        triggers,
        scope,
    };

    Ok(SkillDefinition {
        meta,
        path: path.to_path_buf(),
        body: body.trim().to_string(),
    })
}

/// Split a Markdown file into YAML frontmatter fields and body text.
fn split_frontmatter(text: &str) -> (serde_json::Map<String, serde_json::Value>, String) {
    let lines: Vec<&str> = text.lines().collect();

    if lines.is_empty() || lines[0].trim() != "---" {
        return (serde_json::Map::new(), text.to_string());
    }

    // Find closing ---
    let end_index = lines[1..]
        .iter()
        .position(|line| line.trim() == "---");

    let (frontmatter_text, body) = match end_index {
        Some(i) => {
            let end = i + 1; // 1-based offset from the opening ---
            let fm = lines[1..end].join("\n");
            let body = lines[end + 1..].join("\n");
            (fm, body)
        }
        None => return (serde_json::Map::new(), text.to_string()),
    };

    // Parse YAML-like frontmatter as simple key: value pairs
    let fields = parse_simple_frontmatter(&frontmatter_text);
    (fields, body)
}

/// Parse simple YAML frontmatter (key: value pairs, lists).
/// We use a minimal parser rather than pulling in a full YAML crate.
fn parse_simple_frontmatter(text: &str) -> serde_json::Map<String, serde_json::Value> {
    let mut map = serde_json::Map::new();
    let mut current_list_key: Option<String> = None;
    let mut current_list: Vec<serde_json::Value> = Vec::new();

    for line in text.lines() {
        let trimmed = line.trim();

        // List item: "- value"
        if trimmed.starts_with("- ") && current_list_key.is_some() {
            let value = trimmed[2..].trim().trim_matches('"').to_string();
            current_list.push(serde_json::Value::String(value));
            continue;
        }

        // Flush any pending list
        if let Some(key) = current_list_key.take() {
            if !current_list.is_empty() {
                map.insert(key, serde_json::Value::Array(current_list));
            }
            current_list = Vec::new();
        }

        // Key: value
        if let Some(colon_pos) = trimmed.find(':') {
            let key = trimmed[..colon_pos].trim().to_string();
            let value = trimmed[colon_pos + 1..].trim().to_string();

            if value.is_empty() {
                // Start of a list
                current_list_key = Some(key);
                current_list = Vec::new();
            } else {
                // Simple key: value
                let json_val = if value == "true" {
                    serde_json::Value::Bool(true)
                } else if value == "false" {
                    serde_json::Value::Bool(false)
                } else {
                    serde_json::Value::String(value.trim_matches('"').to_string())
                };
                map.insert(key, json_val);
            }
        }
    }

    // Flush final list
    if let Some(key) = current_list_key.take() {
        if !current_list.is_empty() {
            map.insert(key, serde_json::Value::Array(current_list));
        }
    }

    map
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_skill_name() {
        assert_eq!(normalize_skill_name("  TDD  "), "tdd");
        assert_eq!(normalize_skill_name("Systematic-Debugging"), "systematic-debugging");
    }

    #[test]
    fn test_parse_simple_frontmatter() {
        let text = "name: test-skill\ndescription: A test\nenabled: true";
        let fields = parse_simple_frontmatter(text);
        assert_eq!(fields["name"], "test-skill");
        assert_eq!(fields["description"], "A test");
        assert_eq!(fields["enabled"], true);
    }

    #[test]
    fn test_parse_frontmatter_with_list() {
        let text = "name: test\ntriggers:\n- debug\n- fix";
        let fields = parse_simple_frontmatter(text);
        let triggers = fields["triggers"].as_array().unwrap();
        assert_eq!(triggers.len(), 2);
        assert_eq!(triggers[0], "debug");
        assert_eq!(triggers[1], "fix");
    }

    #[test]
    fn test_split_frontmatter_no_frontmatter() {
        let text = "Just a body\nNo frontmatter";
        let (fields, body) = split_frontmatter(text);
        assert!(fields.is_empty());
        assert_eq!(body, text);
    }

    #[test]
    fn test_split_frontmatter_with_frontmatter() {
        let text = "---\nname: my-skill\n---\nBody content here";
        let (fields, body) = split_frontmatter(text);
        assert_eq!(fields["name"], "my-skill");
        assert!(body.contains("Body content here"));
    }
}
